src/glyph/pen.rs

```diff
@@ -0,0 +1,532 @@
+        match &mut self.scratch_contour {
+            Some(c) => {
+                match point.typ {
+                    PointType::Move => {
+                        if !c.points.is_empty() {
+                            return Err(ErrorKind::UnexpectedMove);
+                        }
+                    }
+                    PointType::Line => {
+                        if self.number_of_offcurves > 0 {
+                            return Err(ErrorKind::UnexpectedPointAfterOffCurve);
+                        }
+                    }
+                    PointType::OffCurve => self.number_of_offcurves += 1,
+                    PointType::QCurve => self.number_of_offcurves = 0,
+                    PointType::Curve => {
+                        if self.number_of_offcurves > 2 {
+                            return Err(ErrorKind::TooManyOffCurves);
+                        }
+                        self.number_of_offcurves = 0;
+                    }
+                }
+                if let Some(identifier) = &point.identifier {
+                    // TODO: test membership at fn start and insert() before push()?
+                    let identifier_hash = Self::hash_identifier(&identifier);
+                    if !self.identifiers.insert(identifier_hash) {
+                        return Err(ErrorKind::DuplicateIdentifier);
+                    }
+                }
+                c.points.push(point);
+                Ok(self)
+            }
+            None => Err(ErrorKind::PenPathNotStarted),
+        }
+    }
+
+    // XXX: should scratch_contour stay in case of error?
+    pub fn end_path(&mut self) -> Result<&mut Self, ErrorKind> {
+        // If we are a closed contour and ended with off-curves, loop around and continue checking from the start.
+        let contour = self.scratch_contour.take();
+        match contour {
+            Some(mut c) => {
+                // XXX: converting here will always leave behind an empty outline,
+                // even there is just an anchor?
+                if Self::contour_is_v1_anchor(&self.glyph.format, &c.points) {
+                    let anchor_point = c.points.remove(0);
+                    let anchor = Anchor {
+                        name: anchor_point.name,
+                        x: anchor_point.x,
+                        y: anchor_point.y,
+                        identifier: None,
+                        color: None,
+                    };
+                    self.glyph.anchors.get_or_insert(Vec::new()).push(anchor);
+                    Ok(self)
+                } else {
+                    if self.number_of_offcurves > 0 {
+                        if c.is_closed() {
+                            for point in &c.points {
+                                match point.typ {
+                                    PointType::OffCurve => self.number_of_offcurves += 1,
+                                    PointType::QCurve => break,
```

Comment:
should we not also check for the number of offcurves?

Reply:
You can have as many off-curves in front of a qcurve as you want I think.

Reply:
ah I'd been thinking of `qcurve` is just a quadratic. The docs aren't clear on this point, but what would it mean to have a qcurve after, say, 4 off-curve points?

Reply:
I think you can have infinite off-curves before a qcurve, they are basically compressed qcurves with an implicit qcurve between every off-curve.